neo/core/container.py

```diff
@@ -73,25 +75,132 @@ def filterdata(data, targdict=None, objects=None, **kwargs):
         results = []
         for key, value in sorted(targdict.items()):
             for obj in data:
-                if (hasattr(obj, key) and getattr(obj, key) == value and
-                        all([obj is not res for res in results])):
+                if (hasattr(obj, key) and getattr(obj, key) == value and all(
+                        [obj is not res for res in results])):
                     results.append(obj)
-                elif (key in obj.annotations and obj.annotations[key] == value and
-                          all([obj is not res for res in results])):
+                elif (isinstance(value, FilterCondition)):
+                    if (key in obj.annotations and value.test(obj.annotations[key]) and all(
+                            [obj is not res for res in results])):
+                        results.append(obj)
+                elif key in obj.annotations and obj.annotations[key] == value and all(
+                        [obj is not res for res in results]):
                     results.append(obj)
 
     # keep only objects of the correct classes
     if objects:
         results = [result for result in results if
-                   result.__class__ in objects or
-                   result.__class__.__name__ in objects]
+                   result.__class__ in objects or result.__class__.__name__ in objects]
 
     if results and all(isinstance(obj, SpikeTrain) for obj in results):
         return SpikeTrainList(results)
     else:
         return results
 
 
+class FilterCondition():
+    """
+        FilterCondition object is given as parameter to container.filter():
+
+        example.filter(annotation=<FilterCondition>) or
+        example=filter({'annotation': <FilterCondition>})
+    """
+
+    def __init__(self, z):
+        pass
+
+    def test(self, x):
+        return True
```

Comment:
You expect this to be overwritten by each subclass, in that case wouldn't it be better to do:
```suggestion
    def test(self, x):
        raise NotImplementedError()
```